File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        stack<int>x;
        int i=0,res=0;
        x.push(-1);
        while(i<s.length()){
            if(s[i]=='(') x.push(i);
            else{
                if(!x.empty()) x.pop();
                if(!x.empty()) {
                    res=max(res,i-x.top());
                }
                else x.push(i);
            }
            i++;
        }
        return res;
    }
};

File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        st=[-1]
        res = 0
        for i in range(len(s)):
            if s[i]=='(':
                st.append(i)
            else:
                st.pop()
                if len(st)==0:
                    st.append(i)
                else:
                    res=max(res,i-st[-1])
        return res
